src/main/java/uk/gov/justice/digital/hmpps/keyworker/controllers/KeyworkerServiceController.java

```diff
@@ -400,20 +400,25 @@ public void migrate(@PathVariable("agencyId") String agencyId) {
             @ApiResponse(code = 400, message = "Invalid request.", response = ErrorResponse.class),
             @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class) })
 
-    @PostMapping(path = "/{staffId}/keyworker")
-    @PreAuthorize("#oauth2.hasScope('admin')")
+    @PostMapping(path = "/{staffId}/agencyId/{agencyId}")
     public void addOrUpdateKeyworker(
             @ApiParam(value = "staffId", required = true)
             @NotEmpty
             @PathVariable("staffId")
-                    long staffId,
+                    Long staffId,
+
+            @ApiParam(value = "agencyId", required = true)
+            @NotEmpty
+            @PathVariable("agencyId")
+                    String agencyId,
+
 
             @ApiParam(value = "New keyworker details." , required=true )
             @Valid
             @RequestBody
                     KeyworkerUpdateDto keyworkerUpdateDto
 
     ) {
-        keyworkerService.addOrUpdate(staffId, keyworkerUpdateDto);
+        keyworkerService.addOrUpdate(staffId, agencyId, keyworkerUpdateDto);
```

Comment:
require agency id as potential deallocations must be restricted to current agency